### PR TITLE
Add shorthand methods for the various status classes in StatusLine

### PR DESCRIPTION
**Pull request by larshelge**
This PR adds shorthand convenience methods for the various status classes from `StatusClass` in `StatusLine`. The purpose is to enable a more concise and shorter notation for checking the class of the HTTP response status.

The main use-case is to easily detect whether a response is a client-side or server-side error.

Currently you have to compare the status class with the inner `StatusClass` enum which is slightly awkward:

```java
if (statusLine.getStatusClass() == StatusLine.StatusClass.CLIENT_ERROR) {
    // Throw a client side exception
}
```

With this patch, the check is more concise and expressive:

```java
if (statusLine.isClientError()) {
    // Throw a client side exception
}
```

It also encapsulates status classes which imply errors with:

```java
if (statusLine.isError()) {
    // Throw an exception
}
```

This will be equivalent to Spring's [HttpStatus](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/http/HttpStatus.html#is1xxInformational--).

### CODE REVIEW

**Review by michael-o — CHANGES_REQUESTED** [review]
I consider the `Yxx` in the method names redundant because status classes are welldefined.

**Comment by larshelge** [comment]
I can agree with that @michael-o . I have updated the PR now. Thanks for the quick response.

**Review by michael-o — APPROVED** [review]
Makes sense and reduces client boilerplate code.

**Comment by garydgregory** [comment]
It seems to me that all the new methods should have the same prefix, not just "is" but maybe "isStatus". Also the public methods should have "@since" Javadoc tags.

**Comment by michael-o** [comment]
@garydgregory Why? The class name says `StatusLine`. Does it make sense to duplicate part of that name into the method names?

**Comment by larshelge** [comment]
On one side there is some merit in having a common prefix to indicate those six methods being related; on the other side it will lead to some duplication/redundancy in naming. You decide.

**Review by ok2c — APPROVED** [review]

**Comment by garydgregory** [comment]
> On one side there is some merit in having a common prefix to indicate those six methods being related; on the other side it will lead to some duplication/redundancy in naming and the "status" part is somewhat implied in the class name (as mentioned).
> 
> Added @ since tags now.

I would really prefer a common prefix to indicate that all of these methods are not only related but mutually exclusive, IOW only one isStatus will return true. The common prefix removed any guesswork IMO.

**Comment by michael-o** [comment]
> > On one side there is some merit in having a common prefix to indicate those six methods being related; on the other side it will lead to some duplication/redundancy in naming and the "status" part is somewhat implied in the class name (as mentioned).
> > Added @ since tags now.
> 
> I would really prefer a common prefix to indicate that all of these methods are not only related but mutually exclusive, IOW only one isStatus will return true. The common prefix removed any guesswork IMO.

Why can't we simply document this?

**Comment by carterkozak** [comment]
A common prefix would be similar to adding these methods directly to `StatusClass`, for example

```java
// original:
statusLine.isError();
// prefix proposal
// At a glance this looks like we're checking for malformed status lines
statusLine.isStatusError();
// alternative StatusClass utility functionality
statusLine.getStatusClass().isError();
```

Usage wouldn't be quite as concise as the current proposal, but it might be a better fit than extending method names.

That said, I don't have a preference one way or the other, just pointing out an alternative. The utility functionality will be helpful.

**Comment by garydgregory** [comment]
You are correct, I lost sight that this code is in the status line class, I thought it was in a class that held a status line instance. Please disregard!

File: httpcore5/src/main/java/org/apache/hc/core5/http/message/StatusLine.java [hunk 1]
```diff
@@ -100,6 +99,61 @@ public StatusClass getStatusClass() {
...
+    }
+
+    /**
+     * Whether this status code is in the HTTP series {@link StatusClass#CLIENT_ERROR}
+     * or {@link StatusClass#SERVER_ERROR}.
+     *
+     * @since 5.1
+     */
+    public boolean isError() {
+        return (isClientError() || isServerError());
```
Comment (garydgregory) [inline httpcore5/src/main/java/org/apache/hc/core5/http/message/StatusLine.java:154]: These extra parents are not needed.

Comment (larshelge) [reply httpcore5/src/main/java/org/apache/hc/core5/http/message/StatusLine.java]: Fixed now.

**Comment by michael-o** [comment]
Looks fine for me now.

**Pull request merged**